server/service/hosts.go

```diff
@@ -238,6 +238,88 @@ func (svc *Service) countHostFromFilters(ctx context.Context, labelID *uint, opt
 	return count, nil
 }
 
+/////////////////////////////////////////////////////////////////////////////////
+// Search
+/////////////////////////////////////////////////////////////////////////////////
+
+type searchHostsRequest struct {
+	// MatchQuery is the query SQL
+	MatchQuery string `json:"query"`
+	// QueryID is the ID of a saved query to run (used to determine if this is a
+	// query that observers can run).
+	QueryID *uint `json:"query_id"`
+	// ExcludedHostIDs is the list of IDs selected on the caller side
+	// (e.g. the UI) that will be excluded from the returned payload.
+	ExcludedHostIDs []uint `json:"excluded_host_ids"`
+}
+
+type searchHostsResponse struct {
+	Hosts []*hostSearchResult `json:"hosts"`
+	Err   error               `json:"error,omitempty"`
+}
+
+func (r searchHostsResponse) error() error { return r.Err }
+
+func searchHostsEndpoint(ctx context.Context, request interface{}, svc fleet.Service) (interface{}, error) {
+	req := request.(*searchHostsRequest)
+
+	hosts, err := svc.SearchHosts(ctx, req.MatchQuery, req.QueryID, req.ExcludedHostIDs)
+	if err != nil {
+		return searchHostsResponse{Err: err}, nil
+	}
+
+	results := []*hostSearchResult{}
+
+	for _, h := range hosts {
+		results = append(results,
+			&hostSearchResult{
+				HostResponse{
+					Host:   h,
+					Status: h.Status(time.Now()),
+				},
+				h.Hostname,
+			},
+		)
+	}
+
+	return searchHostsResponse{
+		Hosts: results,
+	}, nil
+}
+
+func (svc *Service) SearchHosts(ctx context.Context, matchQuery string, queryID *uint, excludedHostIDs []uint) ([]*fleet.Host, error) {
+	if err := svc.authz.Authorize(ctx, &fleet.Host{}, fleet.ActionRead); err != nil {
+		return nil, err
+	}
+
+	vc, ok := viewer.FromContext(ctx)
+	if !ok {
+		return nil, fleet.ErrNoContext
+	}
+
+	includeObserver := false
+	if queryID != nil {
+		query, err := svc.ds.Query(ctx, *queryID)
```

Comment:
FYI this loads packs for a query and does a few other things. All of them are likely fast, but if the goal is to only get `ObserverCanRun`, we can add a ds func that returns that for a query id in a much simpler way.

It might be an over optimization given how often SearchHosts will be called (I think).